vortex-file/src/generic.rs

```diff
@@ -1,677 +1,155 @@
-}
-
-#[derive(Debug)]
-struct CoalescedSegmentRequest {
-    /// The alignment of the first segment.
-    // TODO(ngates): is this the best alignment to use?
-    pub(crate) alignment: Alignment,
-    /// The range of the file to read.
-    pub(crate) byte_range: Range<u64>,
-    /// The original segment requests, ordered by segment ID.
-    pub(crate) requests: Vec<(SegmentId, SegmentSpec)>,
-}
-
-#[derive(Default)]
-struct CoalescedCancellationHandle {
-    handles: Vec<oneshot::Receiver<()>>,
-    cancel_received: Arc<Counter>,
-    cancelled: Arc<Counter>,
-}
-
-impl CoalescedCancellationHandle {
-    fn new(
-        handles: Vec<oneshot::Receiver<()>>,
-        cancel_received: Arc<Counter>,
-        cancelled: Arc<Counter>,
-    ) -> Self {
-        Self {
-            handles,
-            cancel_received,
-            cancelled,
-        }
-    }
-
-    fn push(&mut self, handle: oneshot::Receiver<()>) {
-        self.handles.push(handle);
-    }
-
-    async fn cancelled(self) {
-        for rx in self.handles {
-            // if this segment is completed before cancellation,
-            // tx of this will be dropped, so ignore errors.
-            let _ = rx.await;
-            self.cancel_received.inc();
-        }
-        self.cancelled.inc();
-    }
+struct GenericVortexFileIo<R> {
+    read: Mutex<R>,
+    footer: Footer,
+    segment_cache: Arc<dyn SegmentCache>,
+    options: GenericFileOptions,
 }
 
-async fn evaluate<R: VortexReadAt>(
-    read: R,
-    request: CoalescedSegmentRequest,
-    segment_map: Arc<[SegmentSpec]>,
-    segment_cache: Arc<dyn SegmentCache>,
-    inflight_segments: InflightSegments,
-) -> VortexResult<()> {
-    log::debug!(
-        "Reading byte range for {} requests {:?} size={}",
-        request.requests.len(),
-        request.byte_range,
-        request.byte_range.end - request.byte_range.start,
-    );
-    let buffer: ByteBuffer = read
-        .read_byte_range(request.byte_range.clone(), request.alignment)
-        .await?
-        .aligned(Alignment::none());
+impl<R: VortexReadAt + Send> VortexFileIo for GenericVortexFileIo<R> {
+    fn segment_source(&self, metrics: VortexMetrics) -> Arc<dyn SegmentSource> {
+        // We use segment events for driving I/O.
+        let (segment_source, events) = SegmentEvents::create();
 
-    // Figure out the segments covered by the read.
-    let start = segment_map.partition_point(|s| s.offset < request.byte_range.start);
-    let end = segment_map.partition_point(|s| s.offset < request.byte_range.end);
+        // Wrap the source to resolve segments from the initial read cache.
+        let segment_source = Arc::new(CachedSegmentSource::new(
+            self.segment_cache.clone(),
+            segment_source,
+        ));
 
-    // Note that we may have multiple requests for the same segment.
-    let mut requests_iter = request.requests.into_iter().peekable();
+        let read = self.read.lock().vortex_expect("poisoned lock").clone();
```

Comment:
nit - the `parking_lot` locks are really much nicer, they just don't return a result (and should have better performance).